### ISSUE TITLE
ClientDistributedObjectListenerTest.testDestroyEventReceived_WhenDestroyedFromDifferentInstance test failure

### ISSUE THREAD

**Issue by lazerion** [hazelcast/hazelcast#16396]
- Fails on `Hazelcast-3.x-OracleJDK9` `Hazelcast-3.x-Solaris-OracleJDK8` `Hazelcast-3.x-ZuluJDK12`
- In [this job](http://jenkins.hazelcast.com/view/Official%20Builds/job/Hazelcast-3.x-OracleJDK9/55/testReport/junit/com.hazelcast.client.impl.proxy/ClientDistributedObjectListenerTest/testDestroyEventReceived_WhenDestroyedFromDifferentInstance/)
- Stack trace
```
java.lang.AssertionError: expected:<1> but was:<2>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:631)
	at com.hazelcast.core.DistributedObjectListenerTest.lambda$testDestroyEventReceived_WhenDestroyedFromDifferentInstance$2(DistributedObjectListenerTest.java:144)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1356)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1458)
	at com.hazelcast.core.DistributedObjectListenerTest.testDestroyEventReceived_WhenDestroyedFromDifferentInstance(DistributedObjectListenerTest.java:152)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:564)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:114)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.lang.Thread.run(Thread.java:844)
```

**Comment by ihsandemir** [comment]
duplicate issue. Same problems as in https://github.com/hazelcast/hazelcast/issues/16374 and the test is fixed so that it should not fail anymore. Closing the issue, please re-open if it still fails after fix PR https://github.com/hazelcast/hazelcast/pull/16379

**Comment by lazerion** [comment]
Failed on 01/07 today on [Hazelcast-3.x-Windows-OracleJDK8](http://jenkins.hazelcast.com/view/Official%20Builds/job/Hazelcast-3.x-Windows-OracleJDK8/lastCompletedBuild/testReport/com.hazelcast.client.impl.proxy/ClientDistributedObjectListenerTest/testDestroyEventReceived_WhenDestroyedFromDifferentInstance/)

**Comment by ihsandemir** [comment]
Merged a [fix](https://github.com/hazelcast/hazelcast/pull/16418) to the test hoping that this issue is solved.

**Comment by lazerion** [comment]
Failed on 01/11 on [Hazelcast-3.x-Windows-OracleJDK8 build-132](http://jenkins.hazelcast.com/view/Official%20Builds/job/Hazelcast-3.x-Windows-OracleJDK8/132/testReport/junit/com.hazelcast.client.impl.proxy/ClientDistributedObjectListenerTest/testDestroyEventReceived_WhenDestroyedFromDifferentInstance/)

kindly check

**Comment by ahmetmircik** [comment]
Another failure:
http://jenkins.hazelcast.com/job/Hazelcast-pr-builder/5601/
```
Error Message
expected:<1> but was:<0>
Stacktrace
java.lang.AssertionError: expected:<1> but was:<0>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:631)
	at com.hazelcast.core.DistributedObjectListenerTest.lambda$getVerifyProxyDestroyedTask$1(DistributedObjectListenerTest.java:222)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1356)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:1458)
	at com.hazelcast.core.DistributedObjectListenerTest.verifyDestroy(DistributedObjectListenerTest.java:215)
	at com.hazelcast.core.DistributedObjectListenerTest.testDestroyEventReceived_WhenDestroyedFromDifferentInstance(DistributedObjectListenerTest.java:154)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:114)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
```

**Comment by ihsandemir** [comment]
I examined the [logs](http://jenkins.hazelcast.com/job/Hazelcast-pr-builder/5601/) for the failure and found out this:

![IMG_20200123_134209](https://user-images.githubusercontent.com/7543261/72979749-114b2380-3dea-11ea-8714-eb6b795fd7ce.jpg)

1. There are 3 members and 2 clients. client+16 registers the listener. client_19 destroys the map.
2. Client 19 destroys the map (after line 509) before the client_16 completed registrations to all members (This is line 645).

Both clients have registrations to distributed object events via proxy service start but client_16 also registers the second user listener in the test hence in the logs, client_16 has 6 lines (3 members * 2 listeners) for listener registrations.

We checked the client start code and determined that the listener registrations both internally and config registered ones, may be registered before the client is connected to all members in the cluster which may cause the listener registrations to be late and miss events at the beginning. I will submit a PR for this.
